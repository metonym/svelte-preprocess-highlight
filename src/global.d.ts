import type { PreprocessorGroup } from "svelte/types/compiler/preprocess";

declare global {
  type PreprocessorType = keyof PreprocessorGroup;

  interface SveltePreprocessor<PreprocessorType, Options = any> {
    (options?: Options): Required<Pick<PreprocessorGroup, PreprocessorType>>;
  }

  interface AstNode {
    start: number;
    end: number;
    name: string;
    attributes: Array<{
      name: string;
      value: true | [{ raw: string }];
    }>;
    children: [
      | undefined
      | {
          type: "MustacheTag";
          expression?: {
            value?: string;
            quasis?: [
              {
                value: {
                  raw: string;
                };
              }
            ];
          };
        }
    ];
  }
}
