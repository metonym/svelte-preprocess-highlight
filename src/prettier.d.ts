declare module "prettier-plugin-svelte";

interface PrettierError {
  codeFrame: string;
}
