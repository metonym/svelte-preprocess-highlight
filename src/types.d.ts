export interface AstNode {
  start: number;
  end: number;
  name: string;
  attributes: Array<{
    name: string;
    value: true | [{ raw: string }];
  }>;
  children: [
    | undefined
    | {
        type: "MustacheTag";
        expression?: {
          value?: string;
          quasis?: [
            {
              value: {
                raw: string;
              };
            }
          ];
        };
      }
  ];
}