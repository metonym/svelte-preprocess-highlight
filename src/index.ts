import hljs from "highlight.js";
import MagicString from "magic-string";
import path from "path";
import type { Options as PrettierOptions } from "prettier";
import { format } from "prettier";
import "prettier-plugin-svelte";
import { parse, walk } from "svelte/compiler";
import type { SveltePreprocessor } from "svelte/types/compiler/preprocess";
import type { AstNode } from "./types";

const languages = new Set([...hljs.listLanguages(), "svelte", "html", "auto"]);

interface HighlightOptions {
  /**
   * @default
   * (filename) => /(node_modules|.svelte-kit)/.test(filename)
   * @example
   * highlight({
   *  ignorePath: (filename) => /demo/.test(filename),
   * })
   */
  ignorePath?: (filename: string) => boolean;

  /**
   * Code is formatted by Prettier before it's highlighted.
   * By default, the language is used as the parser.
   * @example
   * highlight({
   *  prettierOptions: {
   *    printWidth: 100,
   *  },
   * })
   */
  prettierOptions?: PrettierOptions;
}

export const highlight: SveltePreprocessor<"markup", HighlightOptions> = (options) => {
  const ignorePath =
    options?.ignorePath || ((filename) => /(node_modules|.svelte-kit)/.test(filename));
  const prettierOptions = options?.prettierOptions;

  return {
    markup({ content, filename }) {
      if (!filename) return;
      if (ignorePath(filename)) return;
      if (!content) return;

      const s = new MagicString(content);
      const log = (...message: unknown[]) => {
        const file = path.relative(process.cwd(), filename);
        console.log(`[${file}]`, ...message);
      };

      walk(parse(content), {
        enter(node: AstNode) {
          if (node.name === "pre") {
            let code = "";
            let language: undefined | string;

            const { attributes, children } = node;
            const expression = children.find((child) => child?.type === "MustacheTag")?.expression;

            if (expression?.value) {
              /**
               * @example
               * <pre data-language="typescript">
               *   {'const a = (b: number) => 4'}
               * </pre>
               */
              code = expression.value;
            } else if (expression?.quasis) {
              /**
               * @example
               * <pre data-language="typescript">
               *   {\`const a = (b: number) => 4\`}
               * </pre>
               */
              code = expression.quasis[0]?.value.raw;
            }

            const language_value = attributes.find(({ name }) => name === "data-language")?.value;

            if (language_value === true) {
              language = "auto";
            } else {
              language = language_value?.[0]?.raw;
            }

            /**
             * Assume normal usage of the `<pre>` element
             * if no `data-language="<language>"` is provided.
             */
            if (!language) return;

            if (!code) return log("No code provided");

            if (!languages.has(language)) return log(`Invalid language "${language}"`);

            let formatted = code;

            try {
              formatted = format(code, { parser: language, ...prettierOptions });
            } catch (error) {
              log("Formatting error", (error as PrettierError).codeFrame);
            }

            let highlighted = formatted;

            if (language === "auto") {
              const auto_highlighted = hljs.highlightAuto(formatted);

              highlighted = auto_highlighted.value;
              language = auto_highlighted.language;
            } else {
              if (/svelte|html/.test(language)) {
                highlighted = hljs.highlightAuto(formatted, ["xml", "css", "javascript"]).value;
              } else {
                highlighted = hljs.highlight(formatted, { language }).value;
              }
            }

            s.overwrite(
              node.start,
              node.end,
              `<pre><code class="hljs">{@html \`${highlighted}\`}</code></pre>`
            );
          }
        },
      });

      return {
        code: s.toString(),
        map: s.generateMap({ source: filename, includeContent: true }),
      };
    },
  };
};
