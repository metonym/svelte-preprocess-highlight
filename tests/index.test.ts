import type { Processed } from "svelte/types/compiler/preprocess";
import { afterEach, describe, expect, test, vi } from "vitest";
import * as API from "../src";

describe("svelte-preprocess-highlight", () => {
  const consoleLog = vi.spyOn(console, "log");

  afterEach(() => {
    consoleLog.mockReset();
  });

  test("API", () => {
    expect(Object.keys(API)).toMatchInlineSnapshot(`
      [
        "highlight",
      ]
    `);
  });

  const p = (
    content: string,
    filename: string = "App.svelte",
    options?: Parameters<typeof API.highlight>[0]
  ) => (API.highlight(options).markup?.({ content, filename }) as undefined | Processed)?.code;

  const TYPESCRIPT_MESSY = `<pre data-language="typescript">{'const a= (b:number) =>4'}</pre>`;
  const TYPESCRIPT_STRING = `<pre data-language="typescript">{'const a = (b: number) => 4'}</pre>`;
  const TYPESCRIPT_QUASIS = `<pre data-language="typescript">{\`const a = (b: number) => 4\`}</pre>`;
  const TYPESCRIPT_QUASIS_LINE_BREAKS = `<pre data-language="typescript">
  
  {\`const a = (b: number) => 4\`}
  </pre>`;
  const TYPESCRIPT_NO_CODE = `<pre data-language="typescript" />`;
  const TYPESCRIPT_NO_LANGUAGE = `<pre>{'const a = (b: number) => 4'}</pre>`;
  const TYPESCRIPT_FORMATTING_ERROR = `<pre data-language="typescript">{'const a= b:number) =>4'}</pre>`;
  const INVALID_LANG = `<pre data-language="hypescript">{'hype'}</pre>`;
  const SVELTE = `<pre data-language="svelte">{\`
  <script>
    let count = 0;
  <\/script>
  
  {@html count}
  \`}
  </pre>`;
  const HTML = `<pre data-language="html">{\`
  <script>
    let count = 0;
  <\/script>
  
  {@html count}
  \`}
  </pre>`;
  const TRIM_LINE_BREAKS = `<pre data-language="typescript">{\`
  const a = (b: number) => 4
  
  
  \`}</pre>`;
  const SVELTE_MULTILINE = `<pre data-language="svelte">{\`
  <script>
    import Highlight from "svelte-highlight";
    import typescript from "svelte-highlight/languages/typescript";
    import atomOneDark from "svelte-highlight/styles/atom-one-dark";
  
    const code = "const add = (a: number, b: number) => a + b;";
  </script>
  
  <svelte:head>
    {@html atomOneDark}
  </svelte:head>
  
  <Highlight language={typescript} {code} />\`}</pre>`;

  const AUTO_HIGHLIGHT = `<pre data-language>{\`body { color: red; }\`}</pre>`;

  const AUTO_HIGHLIGHT_EXPLICIT = `<pre data-language="auto">{\`body { color: red; }\`}</pre>`;

  test("filter – no file name", () => {
    expect(p("content", "")).toEqual(undefined);
  });

  test("filter – skipped file", () => {
    expect(p("content", "/node_modules/App.svelte")).toEqual(undefined);
    expect(p("content", "/.svelte-kit/App.svelte")).toEqual(undefined);
  });

  test("filter – no content", () => {
    expect(p("")).toEqual(undefined);
  });

  test("filter – custom ignore path", () => {
    const options: Parameters<typeof p>[2] = {
      ignorePath: (filename) => /demo/.test(filename),
    };

    expect(p("content", "/node_modules/App.svelte", options)).toEqual("content");
    expect(p("content", "/.svelte-kit/App.svelte", options)).toEqual("content");
    expect(p("content", "/demo/App.svelte", options)).toEqual(undefined);
  });

  test("typescript – mustache tag (string, formatted)", () => {
    expect(p(TYPESCRIPT_MESSY)).toEqual(p(TYPESCRIPT_STRING));
  });

  test("typescript – mustache tag (string)", () => {
    expect(p(TYPESCRIPT_STRING)).toMatchSnapshot();
  });

  test("typescript – mustache tag (quasis)", () => {
    expect(p(TYPESCRIPT_QUASIS)).toMatchSnapshot();
  });

  test("typescript – mustache tag (quasis with line breaks)", () => {
    expect(p(TYPESCRIPT_QUASIS_LINE_BREAKS)).toMatchSnapshot();
    expect(p(TYPESCRIPT_QUASIS_LINE_BREAKS)).toEqual(p(TYPESCRIPT_QUASIS));
  });

  test("typescript – no code", () => {
    expect(p(TYPESCRIPT_NO_CODE)).toMatchSnapshot();
    expect(consoleLog).toBeCalledWith("[App.svelte]", "No code provided");
  });

  test("typescript – no language", () => {
    expect(p(TYPESCRIPT_NO_LANGUAGE)).toMatchSnapshot();
  });

  test("typescript – formatting error", () => {
    expect(p(TYPESCRIPT_FORMATTING_ERROR)).toMatchSnapshot();
    expect(consoleLog).toBeCalledTimes(1);
  });

  test("invalid language", () => {
    expect(p(INVALID_LANG)).toMatchSnapshot();
    expect(consoleLog).toBeCalledWith("[App.svelte]", 'Invalid language "hypescript"');
  });

  test("svelte", () => {
    expect(p(SVELTE)).toMatchSnapshot();
  });

  test("svelte is an alias for html", () => {
    expect(p(SVELTE)).toEqual(p(HTML)!.replace(/"html"/, "svelte"));
  });

  test("trim line breaks", () => {
    expect(p(TRIM_LINE_BREAKS)).toEqual(p(TYPESCRIPT_QUASIS));
  });

  test("multiple", () => {
    expect(p(TYPESCRIPT_MESSY + "\n\n" + TYPESCRIPT_STRING)).toMatchSnapshot();
  });

  test("svelte - multiline", () => {
    expect(p(SVELTE_MULTILINE)).toMatchSnapshot();
  });

  test("auto-highlight", () => {
    expect(p(AUTO_HIGHLIGHT)).toMatchSnapshot();
    expect(p(AUTO_HIGHLIGHT)).toEqual(p(AUTO_HIGHLIGHT_EXPLICIT));
  });
});
