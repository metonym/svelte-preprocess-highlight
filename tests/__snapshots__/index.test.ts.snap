// Vitest Snapshot v1

exports[`svelte-preprocess-highlight > auto-highlight 1`] = `"<pre><code class=\\"hljs\\">{@html \`<span class=\\"hljs-selector-tag\\">body</span> { <span class=\\"hljs-attribute\\">color</span>: red; }\`}</code></pre>"`;

exports[`svelte-preprocess-highlight > invalid language 1`] = `"<pre data-language=\\"hypescript\\">{'hype'}</pre>"`;

exports[`svelte-preprocess-highlight > multiple 1`] = `
"<pre><code class=\\"hljs\\">{@html \`<span class=\\"hljs-keyword\\">const</span> <span class=\\"hljs-title function_\\">a</span> = (<span class=\\"hljs-params\\">b: <span class=\\"hljs-built_in\\">number</span></span>) =&gt; <span class=\\"hljs-number\\">4</span>;
\`}</code></pre>

<pre><code class=\\"hljs\\">{@html \`<span class=\\"hljs-keyword\\">const</span> <span class=\\"hljs-title function_\\">a</span> = (<span class=\\"hljs-params\\">b: <span class=\\"hljs-built_in\\">number</span></span>) =&gt; <span class=\\"hljs-number\\">4</span>;
\`}</code></pre>"
`;

exports[`svelte-preprocess-highlight > svelte - multiline 1`] = `
"<pre><code class=\\"hljs\\">{@html \`<span class=\\"hljs-tag\\">&lt;<span class=\\"hljs-name\\">script</span>&gt;</span><span class=\\"language-javascript\\">
  <span class=\\"hljs-keyword\\">import</span> <span class=\\"hljs-title class_\\">Highlight</span> <span class=\\"hljs-keyword\\">from</span> <span class=\\"hljs-string\\">&quot;svelte-highlight&quot;</span>;
  <span class=\\"hljs-keyword\\">import</span> typescript <span class=\\"hljs-keyword\\">from</span> <span class=\\"hljs-string\\">&quot;svelte-highlight/languages/typescript&quot;</span>;
  <span class=\\"hljs-keyword\\">import</span> atomOneDark <span class=\\"hljs-keyword\\">from</span> <span class=\\"hljs-string\\">&quot;svelte-highlight/styles/atom-one-dark&quot;</span>;

  <span class=\\"hljs-keyword\\">const</span> code = <span class=\\"hljs-string\\">&quot;const add = (a: number, b: number) =&gt; a + b;&quot;</span>;
</span><span class=\\"hljs-tag\\">&lt;/<span class=\\"hljs-name\\">script</span>&gt;</span>

<span class=\\"hljs-tag\\">&lt;<span class=\\"hljs-name\\">svelte:head</span>&gt;</span>
  {@html atomOneDark}
<span class=\\"hljs-tag\\">&lt;/<span class=\\"hljs-name\\">svelte:head</span>&gt;</span>

<span class=\\"hljs-tag\\">&lt;<span class=\\"hljs-name\\">Highlight</span> <span class=\\"hljs-attr\\">language</span>=<span class=\\"hljs-string\\">{typescript}</span> {<span class=\\"hljs-attr\\">code</span>} /&gt;</span>
\`}</code></pre>"
`;

exports[`svelte-preprocess-highlight > svelte 1`] = `
"<pre><code class=\\"hljs\\">{@html \`<span class=\\"hljs-tag\\">&lt;<span class=\\"hljs-name\\">script</span>&gt;</span><span class=\\"language-javascript\\">
  <span class=\\"hljs-keyword\\">let</span> count = <span class=\\"hljs-number\\">0</span>;
</span><span class=\\"hljs-tag\\">&lt;/<span class=\\"hljs-name\\">script</span>&gt;</span>

{@html count}
\`}</code></pre>"
`;

exports[`svelte-preprocess-highlight > typescript – formatting error 1`] = `"<pre><code class=\\"hljs\\">{@html \`<span class=\\"hljs-keyword\\">const</span> a= <span class=\\"hljs-attr\\">b</span>:<span class=\\"hljs-built_in\\">number</span>) =&gt;<span class=\\"hljs-number\\">4</span>\`}</code></pre>"`;

exports[`svelte-preprocess-highlight > typescript – mustache tag (quasis with line breaks) 1`] = `
"<pre><code class=\\"hljs\\">{@html \`<span class=\\"hljs-keyword\\">const</span> <span class=\\"hljs-title function_\\">a</span> = (<span class=\\"hljs-params\\">b: <span class=\\"hljs-built_in\\">number</span></span>) =&gt; <span class=\\"hljs-number\\">4</span>;
\`}</code></pre>"
`;

exports[`svelte-preprocess-highlight > typescript – mustache tag (quasis) 1`] = `
"<pre><code class=\\"hljs\\">{@html \`<span class=\\"hljs-keyword\\">const</span> <span class=\\"hljs-title function_\\">a</span> = (<span class=\\"hljs-params\\">b: <span class=\\"hljs-built_in\\">number</span></span>) =&gt; <span class=\\"hljs-number\\">4</span>;
\`}</code></pre>"
`;

exports[`svelte-preprocess-highlight > typescript – mustache tag (string) 1`] = `
"<pre><code class=\\"hljs\\">{@html \`<span class=\\"hljs-keyword\\">const</span> <span class=\\"hljs-title function_\\">a</span> = (<span class=\\"hljs-params\\">b: <span class=\\"hljs-built_in\\">number</span></span>) =&gt; <span class=\\"hljs-number\\">4</span>;
\`}</code></pre>"
`;

exports[`svelte-preprocess-highlight > typescript – no code 1`] = `"<pre data-language=\\"typescript\\" />"`;

exports[`svelte-preprocess-highlight > typescript – no language 1`] = `"<pre>{'const a = (b: number) => 4'}</pre>"`;
